### ISSUE TITLE
Using defaultTo in text is being ignored.

### ISSUE THREAD

**Issue by peterwilli** [tgriesser/knex#2649]
I was too busy to fill out the entire form, so instead I added an example to reproduce.

```js
var knex = require('knex')({
  connection: {
    host: 'localhost',
    user: 'root',
    password: '',
    db: 'db'
  },
  client: 'mariasql',
  useNullAsDefault: true
});

knex.schema.createTable('test', function (table) {
  table.increments();
  table.text('hi').defaultTo("test");
}).then();
```

After running the code above the structure of the new table is:
![selection_066](https://user-images.githubusercontent.com/1212814/41061381-acf44ff2-69d2-11e8-8847-d756f88b0c25.png)

It happens with both MariaSQL and MySQL. The latest version I tried this on is `0.14.6`

**Comment by elhigu** [comment]
Looks like this fails also with mssql. Integration test + fix would be nice.

**Comment by wubzz** [comment]
Seems to only apply to `blob` and `text` fields, rest works. They are explicitly excluded in mysql/mssql (maria inherits from mysql). I wonder why that is..

**Comment by mirkojotic** [comment]
Did a quick Google search and this is what came up:
https://bugs.mysql.com/bug.php?id=21532
Stack overflow post where I found it:
https://stackoverflow.com/questions/3466872/why-cant-a-text-column-have-a-default-value-in-mysql

TLDR;
MySQL (MariaDB) doesn't support comments on BLOB/TEXT fields due to `values are not stored directly in the record buffer used for reading/updating tables. So it is a bit more complex to assign default values for them.`

**Comment by peterwilli** [comment]
@mirkojotic damn. Thanks... Weird limit, though.

**Comment by mirkojotic** [comment]
@peterwilli

Official mysql documentation says it's not possible
https://dev.mysql.com/doc/refman/8.0/en/data-type-defaults.html

But the stack overflow link I've put in the previous comment had a solution that can accomplish this using a trigger. 

phpmyadmin might use something like that. I'm going to let people more experienced with the project make the decision 😉

**Comment by elhigu** [comment]
Trigger solution would require also some magic to at least for column.alter() functionality. At least if defaultTo is set and dialect doesn't support it, an error should be thrown instead of silent ignore. Probably throwing an error for now and adding feature request for trigger implementation would be enough to close this.

**Comment by tgriesser** [comment]
I'd rather not do the trigger implementation, an error or warning would probably suffice here. If folks want to get fancy and do it with a trigger I'd put that beyond the scope of knex.

I'd actually probably say warn over error until we have a better migration API since this is mostly applicable to mysql which also (until 8.0 is mainstream) doesn't have DDL transactions, so throwing would probably lead to bad states more often.

**Comment by rotimi-babalola** [comment]
Hello. Is there a way to set `null` as the default value for `table.text`? I'm using Postgres and my migration file looks like this:

```
exports.up = knex => knex.schema.alterTable('books', (table) => {
  table.text('description').defaultTo(null);
});

exports.down = knex => knex.schema.alterTable('bill_differences', (table) => {
  table.dropColumn('description');
});
```

When I run the migration, I don't see `null` as the default value for that column in my database client - Postico. Is there a way to resolve this issue? Thanks

**Comment by kibertoad** [comment]
@rotimi-babalola And when you make column nullable and insert entry without value, what it defaults to?

**Comment by rotimi-babalola** [comment]
It does not default to anything @kibertoad it is just blank

**Comment by kibertoad** [comment]
@rotimi-babalola So you are saying that you are getting different results when you are explicitly passing null and when you are passing nothing?
Documentation for PostgreSQL states this: "If no default value is declared explicitly, the default value is the null value", so I'm slightly confused that you are not getting null by default.

**Comment by rotimi-babalola** [comment]
Oh I see. Thanks @kibertoad

**Comment by pedro-surf** [comment]
Getting the same issues here on MSSQL.

From which Knex version can we use defaultTo method?

**Comment by louislam** [comment]
FYI, MariaDB is supporting default value for TEXT fields now (as least in 2024). 

> BLOB and TEXT columns can be assigned a [DEFAULT](https://mariadb.com/kb/en/create-table/#default) value. This was not possible prior to [MariaDB 10.2](https://mariadb.com/kb/en/what-is-mariadb-102/).

https://mariadb.com/kb/en/text/

**Comment by louislam** [comment]
I fixed it in my side by overriding the `columnCompiler` and `defaultTo`.

https://github.com/louislam/uptime-kuma/pull/5261/files

More information:

MySQL 8.0 is required and only if the value is written as an expression: https://dev.mysql.com/doc/refman/8.0/en/data-type-defaults.html
MariaDB 10.2 is required: https://mariadb.com/kb/en/text/


So I think this constraint will be removed as of today. I can create a pr for this if maintainers agree with this.

**Comment by mercmobily** [comment]
Thank you for reporting this issue! You've identified a long-standing limitation that is now outdated.

Knex explicitly excludes TEXT and BLOB fields from having default values at lib/dialects/mysql/schema/mysql-columncompiler.js:126-129. This was originally based on MySQL's historical restrictions, but the database landscape has changed significantly since 2018.

As louislam correctly points out, this constraint should now be removed. MySQL 5.6 (EOL: 28 Feb 2021) and MariaDB 10.1 (EOL: 17 Oct 2020) both lacked this feature, but modern versions support it:

- **MySQL 8.0.13+** (Oct 2018): Supports defaults for BLOB/TEXT/JSON/GEOMETRY using expression syntax: `DEFAULT ('value')`
- **MariaDB 10.2+** (2017): Supports defaults for TEXT/BLOB directly without expression syntax
- **MSSQL**: Already works (uses `nvarchar(max)`)

Knex currently tests against MySQL 8.0.40, so the infrastructure is already in place to support this feature.

The fix would need to update the MySQL column compiler's `defaultTo()` method to wrap text/blob defaults in parentheses for expression syntax, similar to what was done for JSON defaults at lines 121-124. The uptime-kuma workaround (referenced in comments) demonstrates a working implementation.

I am marking this as a bug, to give it priority. It will get done once we finish the full triaging in progress right now.
